File: app/src/main/java/de/bikebean/app/ui/utils/sms/load/SmsLoader.java
package de.bikebean.app.ui.utils.sms.load;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.os.AsyncTask;
import android.provider.Telephony;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.lang.ref.WeakReference;

import de.bikebean.app.ui.initialization.Conversation;
import de.bikebean.app.db.sms.Sms;
import de.bikebean.app.ui.main.status.StateViewModel;
import de.bikebean.app.ui.main.status.menu.log.LogViewModel;
import de.bikebean.app.ui.main.status.menu.sms_history.SmsViewModel;

public class SmsLoader extends AsyncTask<String, Void, Void> {

    public interface InboxTraverser {
        void traverseInbox(final @NonNull Cursor inbox);
    }

    private final @NonNull WeakReference<Context> contextWeakReference;
    private final SmsViewModel smsViewModel;
    private final StateViewModel stateViewModel;
    private final LogViewModel logViewModel;

    public SmsLoader(final @NonNull Context context, final SmsViewModel smsViewModel,
                     final StateViewModel stateViewModel, final LogViewModel logViewModel) {
        contextWeakReference = new WeakReference<>(context);
        this.smsViewModel = smsViewModel;
        this.stateViewModel = stateViewModel;
        this.logViewModel = logViewModel;
    }

    @Override
    public Void doInBackground(final @NonNull String... args) {
        final @NonNull String phoneNumber = args[0];

        getCursor(phoneNumber, this::traverseInboxAndClose);

        return null;
    }

    public void loadInitial(final @NonNull String phoneNumber) {
        logViewModel.i("Initial Loading");
        getCursor(phoneNumber, this::traverseInboxAndCloseInitial);
        logViewModel.i("Initial Loading completed");
    }

    private void getCursor(final @NonNull String phoneNumber,
                           final @NonNull InboxTraverser inboxTraverser) {
        final @NonNull String[] argList = { phoneNumber };

        final @Nullable Context context = contextWeakReference.get();
        final @Nullable ContentResolver contentResolver;
        final @Nullable Cursor cursor;

        if (context != null)
            contentResolver = context.getContentResolver();
        else {
            logViewModel.w("Failed to load SMS, context was removed!");
            return;
        }

        if (contentResolver != null)
            cursor = getInbox(contentResolver, argList);
        else {
            logViewModel.w("Failed to load SMS, could not get contentResolver!");
            return;
        }

        if (cursor != null)
            inboxTraverser.traverseInbox(cursor);
        else
            logViewModel.w("Failed to load SMS, could not get cursor");
    }

    private @Nullable Cursor getInbox(final @NonNull ContentResolver contentResolver,
                                      final @NonNull String[] argList) {
        return contentResolver.query(
                Telephony.Sms.Inbox.CONTENT_URI, null,
                "address=?", argList,
                null, null);
    }

    private void traverseInboxAndClose(final @NonNull Cursor inbox) {
        if (inbox.moveToFirst()) {

            for (int i=0; i < inbox.getCount(); i++) {
                final @NonNull Sms sms = new Sms(inbox, Sms.STATUS.NEW);

                if (smsViewModel.getSmsById(sms.getId()).size() == 0)
                    smsViewModel.insert(sms);

                inbox.moveToNext();
            }

            inbox.close();
        }
    }

    private void traverseInboxAndCloseInitial(final @NonNull Cursor inbox) {
        final @NonNull Conversation conversation =
                new Conversation(stateViewModel, smsViewModel, logViewModel);

        if (inbox.moveToFirst()) {
            logViewModel.d("Loading " + inbox.getCount() + " SMS");
            for (int i = 0; i < inbox.getCount(); i++) {
                conversation.add(new Sms(inbox, Sms.STATUS.PARSED));
                inbox.moveToNext();
            }

            logViewModel.d("Done Loading SMS!");
            inbox.close();
        } else
            return;

        conversation.updatePreferences();
    }
}

File: app/src/main/java/de/bikebean/app/ui/utils/sms/load/SmsLoader.kt
package de.bikebean.app.ui.utils.sms.load

import android.content.ContentResolver
import android.content.Context
import android.database.Cursor
import android.provider.Telephony
import de.bikebean.app.db.sms.Sms
import de.bikebean.app.db.sms.SmsFactory
import de.bikebean.app.ui.drawer.log.LogViewModel
import de.bikebean.app.ui.drawer.sms_history.SmsViewModel
import de.bikebean.app.ui.drawer.status.StateViewModel
import de.bikebean.app.ui.initialization.Conversation
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.lang.ref.WeakReference

class SmsLoader(
        private val contextWeakReference: WeakReference<Context?>,
        private val smsViewModel: SmsViewModel?,
        private val stateViewModel: StateViewModel?,
        private val logViewModel: LogViewModel?) {

    constructor(context: Context?,
                smsViewModel: SmsViewModel?,
                stateViewModel: StateViewModel?,
                logViewModel: LogViewModel?) :
            this(WeakReference(context), smsViewModel, stateViewModel, logViewModel)

    constructor(context: Context) :
            this(WeakReference(context),null, null, null)

    fun execute(address: String) {
        CoroutineScope(Dispatchers.Main).launch {
            doInBackground(address)
        }
    }

    private suspend fun doInBackground(address: String) =
            withContext(Dispatchers.IO) {
                getCursor(address)?.let {
                    traverseInboxAndClose(it, ::insertSmsIfNotPresent, null)
                }
            }

    fun loadInitial(phoneNumber: String) {
        val conversation = Conversation(stateViewModel!!, smsViewModel!!, logViewModel!!)

        logViewModel.i("Initial Loading")
        getCursor(phoneNumber)?.let { traverseInboxAndClose(it, conversation::add, conversation) }
        logViewModel.i("Initial Loading completed")
    }

    fun getMessageNumber(phoneNumber: String): Int = getCursor(phoneNumber)?.count ?: 0

    private fun getCursor(phoneNumber: String): Cursor? = contextWeakReference.get()?.let { context ->
        context.contentResolver?.let { contentResolver ->
            getInbox(contentResolver, arrayOf(phoneNumber)) ?: null.also {
                logViewModel?.w("Failed to load SMS, could not get cursor")
            }
        } ?: null.also {
            logViewModel?.w("Failed to load SMS, could not get contentResolver!")
        }
    } ?: null.also {
        logViewModel?.w("Failed to load SMS, context was removed!")
    }

    private fun getInbox(contentResolver: ContentResolver, argList: Array<String>): Cursor? =
            contentResolver.query(
                    Telephony.Sms.Inbox.CONTENT_URI, null,
                    "address=?", argList,
                    null, null
            )

    private fun traverseInboxAndClose(
            inbox: Cursor,
            insertHandler: (Sms) -> Unit,
            conversation: Conversation?) {

        if (inbox.moveToFirst()) {
            conversation?.let { logViewModel?.d("Loading ${inbox.count} SMS") }

            repeat(inbox.count) {
                SmsFactory.createSmsFromCursor(inbox, conversation).let(insertHandler)
                inbox.moveToNext()
            }

            conversation?.let { logViewModel?.d("Done Loading SMS!") }
            inbox.close()
        }

        conversation?.updatePreferences()
    }

    private fun insertSmsIfNotPresent(sms: Sms) {
        smsViewModel ?: return
        when {
            smsViewModel.getSmsById(sms.id).isEmpty() -> sms
            else -> null
        }?.let(smsViewModel::insert)
    }
}